firebase_test.go

```diff
@@ -361,6 +362,44 @@ func TestMessaging(t *testing.T) {
 	}
 }
 
+func TestMessagingSendWithCustomEndpoint(t *testing.T) {
+	name := "custom-endpoint-ok"
+
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.Header().Set("Content-Type", "application/json")
+		w.Write([]byte("{ \"name\":\"" + name + "\" }"))
+	}))
+	defer ts.Close()
+
+	ctx := context.Background()
+
+	tokenSource := &testTokenSource{AccessToken: "mock-token-from-custom"}
+	app, err := NewApp(
+		ctx,
+		nil,
+		option.WithCredentialsFile("testdata/service_account.json"),
+		option.WithTokenSource(tokenSource),
+		option.WithEndpoint(ts.URL),
```

Comment:
Only `WithEndpoint()` should be sufficient here.

Reply:
It gives an error only with `WithEndpoint()`:
```go
app, err := NewApp(ctx, nil, option.WithEndpoint(ts.URL))
if err != nil {
    t.Fatal(err)
}

c, err := app.Messaging(ctx)
if c == nil || err != nil {
    t.Fatalf("Messaging() = (%v, %v); want (iid, nil)", c, err)
}
```
```
--- FAIL: TestMessagingSendWithCustomEndpoint (0.59s)
    firebase_test.go:390: Messaging() = (<nil>, project ID is required to access Firebase Cloud Messaging client); want (iid, nil)
```

Reply:
Ok, I suppose the transport still requires some sort of a credential. In that case let's keep `option.WithTokenSource()`. You definitely don't need `WithCredentialsFile()` when a TokenSource is explicitly provided.

Reply:
It seems `WithCredentialsFile()` is required to provide project id. It gives an error without `WithCredentialsFile()`:

```go
tokenSource := &testTokenSource{AccessToken: "mock-token-from-custom"}
app, err := NewApp(
    ctx,
    nil,
    //option.WithCredentialsFile("testdata/service_account.json"),
    option.WithTokenSource(tokenSource),
    option.WithEndpoint(ts.URL),
)
if err != nil {
    t.Fatal(err)
}

c, err := app.Messaging(ctx)
if c == nil || err != nil {
    t.Fatalf("Messaging() = (%v, %v); want (iid, nil)", c, err)
}
```

```
--- FAIL: TestMessagingSendWithCustomEndpoint (0.00s)
    firebase_test.go:390: Messaging() = (<nil>, project ID is required to access Firebase Cloud Messaging client); want (iid, nil)
```

Reply:
Ah! :(